File: ios/SkyCast/Features/Home/HourlyStrip.swift
import SwiftUI

/// A horizontally scrollable strip of three-hourly readings around the present moment.
///
/// A `ScrollView` supplies momentum, rubber-banding and VoiceOver's scroll actions. Past hours are
/// included, and the present is marked.
///
/// Times are rendered in the **forecast location's** zone, not the device's: a Maldivian forecast
/// read from London must still show Maldivian hours.
struct HourlyStrip: View {
    let hours: [HourlyForecast]
    let timeZone: TimeZone
    let unit: TemperatureUnit
    var now: Date = .now
    /// Overridable because the detail screen shows only what is ahead, where "Through the day"
    /// does not apply.
    var title = "Through the day"

    var body: some View {
        if hours.isEmpty {
            EmptyView()
        } else {
            VStack(alignment: .leading, spacing: Spacing.sm) {
                Text(title)
                    .font(.subheadline.weight(.medium))
                    .foregroundStyle(.secondary)
                    .padding(.horizontal, Spacing.md)

                ScrollView(.horizontal) {
                    HStack(spacing: Spacing.sm) {
                        ForEach(hours) { hour in
                            HourColumn(hour: hour, timeZone: timeZone, unit: unit, now: now)
                        }
                    }
                    .padding(.horizontal, Spacing.md)
                    .scrollTargetLayout()
                }
                .scrollIndicators(.hidden)
            }
        }
    }
}

private struct HourColumn: View {
    let hour: HourlyForecast
    let timeZone: TimeZone
    let unit: TemperatureUnit
    let now: Date

    private var isPast: Bool {
        hour.time < now
    }

    /// Whether this reading covers the present moment. Readings are three hours apart, so "now" is
    /// the most recent one that has already started.
    private var isCurrent: Bool {
        hour.time <= now && now < hour.time.addingTimeInterval(Self.readingInterval)
    }

    private var label: String {
        if isCurrent {
            return "Now"
        }
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = timeZone
        let formatter = DateFormatter()
        formatter.dateFormat = "HH:mm"
        formatter.timeZone = timeZone
        return formatter.string(from: hour.time)
    }

    private var temperature: Int {
        Int(unit.convertFromCelsius(hour.temperatureCelsius).rounded())
    }

    private var rainPercent: Int {
        Int((hour.precipitationProbability * 100).rounded())
    }

    /// The forecast carries no sunrise or sunset, so daylight is approximated by clock hour.
    /// Getting this wrong shows a sun at 3 am, the parity bug `WeatherConditionIconTests` guards.
    private var isDaytime: Bool {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = timeZone
        let hourOfDay = calendar.component(.hour, from: hour.time)
        return hourOfDay >= Self.dawnHour && hourOfDay < Self.duskHour
    }

    private var announcement: String {
        var text = "\(label), \(temperature)\(unit.symbol)"
        if rainPercent > 0 {
            text += ", \(rainPercent) percent chance of rain"
        }
        return text
    }

    var body: some View {
        VStack(spacing: Spacing.xs) {
            Text(label)
                .font(.caption)
                .fontWeight(isCurrent ? .bold : .regular)
                // Past readings recede rather than disappear: still legible, clearly behind us.
                .foregroundStyle(isPast ? AnyShapeStyle(.secondary) : AnyShapeStyle(.primary))

            ConditionBadge(condition: hour.condition, isDaytime: isDaytime, size: badgeSize)

            Text("\(temperature)\(unit.symbol)")
                .font(.subheadline.weight(.semibold))

            // Only when there is something to say. An always-present "0%" is noise on nine days
            // out of ten.
            if rainPercent > 0 {
                Text("\(rainPercent)%")
                    .font(.caption2)
                    .foregroundStyle(WeatherPalette.humidity)
            }
        }
        .frame(width: columnWidth)
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(announcement)
    }

    private let columnWidth: CGFloat = 60
    private let badgeSize: CGFloat = 20
    private static let readingInterval: TimeInterval = 3 * 60 * 60
    private static let dawnHour = 6
    private static let duskHour = 20
}

File: android/app/src/main/java/com/nauhaan/skycast/ui/today/HourlyStrip.kt
package com.nauhaan.skycast.ui.today

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.input.nestedscroll.NestedScrollConnection
import androidx.compose.ui.input.nestedscroll.NestedScrollSource
import androidx.compose.ui.input.nestedscroll.nestedScroll
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Velocity
import androidx.compose.ui.unit.dp
import com.nauhaan.skycast.R
import com.nauhaan.skycast.core.designsystem.component.WeatherConditionBadge
import com.nauhaan.skycast.core.designsystem.theme.Spacing
import com.nauhaan.skycast.core.designsystem.theme.weatherPalette
import com.nauhaan.skycast.domain.model.HourlyForecast
import com.nauhaan.skycast.domain.model.TemperatureUnit
import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import kotlin.math.roundToInt

/**
 * A horizontally scrollable strip of three-hourly readings around the present moment.
 *
 * A `LazyRow` supplies momentum, edge behaviour and accessibility scrolling. Past hours are
 * included, and the present is marked.
 *
 * Times are rendered in the **forecast location's** zone, not the device's: a Maldivian forecast
 * read from London must still show Maldivian hours.
 */
@Composable
fun HourlyStrip(
    hours: List<HourlyForecast>,
    zoneOffset: ZoneOffset,
    unit: TemperatureUnit,
    modifier: Modifier = Modifier,
    now: Instant = Instant.now(),
    // Overridable because the detail screen shows only what is ahead, where "Through the day" does
    // not apply.
    title: String = stringResource(R.string.today_hourly_heading),
) {
    if (hours.isEmpty()) return

    Column(modifier = modifier) {
        Text(
            text = title,
            style = MaterialTheme.typography.titleSmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            modifier = Modifier.padding(start = Spacing.md, bottom = Spacing.sm),
        )
        LazyRow(
            horizontalArrangement = Arrangement.spacedBy(Spacing.sm),
            contentPadding = PaddingValues(horizontal = Spacing.md),
            modifier = Modifier.nestedScroll(StripKeepsItsOwnGesture),
        ) {
            items(hours, key = { it.time.epochSecond }) { hour ->
                HourColumn(hour = hour, zoneOffset = zoneOffset, unit = unit, now = now)
            }
        }
    }
}

/**
 * Stops a scroll of the strip turning into a swipe to the next place.
 *
 * Compose chains nested scrolling: when the row reaches its end, the horizontal delta it could not
 * use travels up to the [HorizontalPager] behind it, and the *same* gesture carries on as a page
 * change.
 *
 * This consumes whatever the row left over, in the horizontal axis only, so the pager sees nothing
 * from a gesture that began here. Vertical is passed through untouched, so a drag up or down on the
 * strip still scrolls the page. Swiping anywhere else still changes place.
 *
 * `internal` rather than private so `HourlyStripGestureTest` can assert both halves.
 */
internal val StripKeepsItsOwnGesture = object : NestedScrollConnection {
    override fun onPostScroll(consumed: Offset, available: Offset, source: NestedScrollSource): Offset =
        available.copy(y = 0f)

    override suspend fun onPostFling(consumed: Velocity, available: Velocity): Velocity = available.copy(y = 0f)
}

@Composable
private fun HourColumn(
    hour: HourlyForecast,
    zoneOffset: ZoneOffset,
    unit: TemperatureUnit,
    now: Instant,
    modifier: Modifier = Modifier,
) {
    val localTime = hour.time.atOffset(zoneOffset)
    val temperature = unit.convertFromCelsius(hour.temperatureCelsius).roundToInt()
    val isCurrent = hour.isCurrent(now)
    // Checked before `isPast`, not after. The reading covering this moment *started* in the past,
    // three-hourly readings always have, so testing "is it past?" first meant the current hour
    // rendered as a time like any other, and only its bold weight hinted otherwise. iOS had the
    // order right, which is how the difference showed up.
    val isPast = hour.time.isBefore(now) && !isCurrent
    val label = if (isCurrent) {
        stringResource(R.string.today_hourly_now)
    } else {
        TIME_FORMAT.format(localTime)
    }
    val rain = (hour.precipitationProbability * PERCENT).roundToInt()

    // The forecast carries no sunrise or sunset, so daylight is approximated by clock hour.
    // Getting this wrong shows a sun at 3 am: the parity bug WeatherConditionIconTest guards.
    val isDaytime = localTime.hour in DAWN_HOUR until DUSK_HOUR

    val announcement = buildString {
        append(stringResource(R.string.today_hourly_accessibility, label, temperature, unit.symbol))
        if (rain > 0) {
            append(", ")
            append(stringResource(R.string.day_detail_rain_accessibility, rain))
        }
    }

    Column(
        modifier = modifier
            .width(ColumnWidth)
            .clearAndSetSemantics { contentDescription = announcement },
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(Spacing.xs),
    ) {
        Text(
            text = label,
            style = MaterialTheme.typography.labelMedium,
            // Past readings recede rather than disappear: still legible, clearly behind us.
            color = if (isPast) {
                MaterialTheme.colorScheme.onSurfaceVariant
            } else {
                MaterialTheme.colorScheme.onSurface
            },
            fontWeight = if (isCurrent) FontWeight.Bold else FontWeight.Normal,
        )
        WeatherConditionBadge(
            condition = hour.condition,
            isDaytime = isDaytime,
            size = BadgeSize,
        )
        Text(
            text = "$temperature${unit.symbol}",
            style = MaterialTheme.typography.titleMedium,
        )
        // Only when there is something to say. An always-present "0%" is noise on nine of ten days.
        if (rain > 0) {
            Text(
                text = "$rain%",
                style = MaterialTheme.typography.labelSmall,
                color = weatherPalette.humidity,
            )
        }
    }
}

/**
 * Whether this reading is the one covering the present moment.
 *
 * Readings are three hours apart, so "now" is the most recent one that has already started.
 */
private fun HourlyForecast.isCurrent(now: Instant): Boolean {
    val start = time
    val end = time.plusSeconds(READING_INTERVAL_SECONDS)
    return !now.isBefore(start) && now.isBefore(end)
}

private val TIME_FORMAT: DateTimeFormatter = DateTimeFormatter.ofPattern("HH:mm")
private val ColumnWidth = 64.dp
private val BadgeSize = 36.dp
private const val PERCENT = 100
private const val DAWN_HOUR = 6
private const val DUSK_HOUR = 20
private const val READING_INTERVAL_SECONDS = 3L * 60 * 60
